### ISSUE TITLE
Bug detected: ! java.lang.ClassCastException: graphql.schema.GraphQLTypeReference cannot be cast to graphql.schema.GraphQLObjectType

### ISSUE THREAD

**Issue by tufgupmi** [leangen/graphql-spqr#349]
Hello, 

When you have for example:
```

@GraphQLUnion(name="Base", possibleAutoDiscovery="true")
class abstract Base {
....
}

class A extends Base {
.....
}

class B extends Base {
.....
}

class Other {
...
   public A getA() {
       return A
  }
}

class Another {
   public Base getResult() {
     return base
  }
}


```

When you call getResult() then is thrown the following exception:
```

ERROR [2020-04-03 16:08:46,821] [dw-33 - POST /graphql] [92078D5572CB49D8AF67CD5035EBCE1B] [LoggingExceptionMapper] Error handling a request: 3d7aeedb2f872ebd 
! java.lang.ClassCastException: graphql.schema.GraphQLTypeReference cannot be cast to graphql.schema.GraphQLObjectType
! at io.leangen.graphql.generator.MappedType.getAsObjectType(MappedType.java:25)
! at io.leangen.graphql.generator.DelegatingTypeResolver.getType(DelegatingTypeResolver.java:51)
! at graphql.execution.ExecutionStrategy.resolveTypeForUnion(ExecutionStrategy.java:761)
! at graphql.execution.ExecutionStrategy.resolveType(ExecutionStrategy.java:721)
! at graphql.execution.ExecutionStrategy.completeValue(ExecutionStrategy.java:412)
! at graphql.execution.ExecutionStrategy.completeField(ExecutionStrategy.java:363)
! at graphql.execution.ExecutionStrategy.lambda$resolveFieldWithInfo$0(ExecutionStrategy.java:201)
! at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
! at java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:614)

```

This is because then resolving the union Base the type returned A is a GraphQLTypeReference and you are trying to cast to GraphQLObjectType...

To solve it i have change the DelegatingTypeResolver.getType
```

 if (mappedTypes.size() == 1) {
            GraphQLOutputType outputType = mappedTypes.get(0).getGraphQLType();
            if (outputType instanceof GraphQLObjectType) {
                return (GraphQLObjectType) outputType;
            } else if (outputType instanceof GraphQLTypeReference) {
                GraphQLTypeReference typeReference = (GraphQLTypeReference) outputType;
                return (GraphQLObjectType) env.getSchema().getTypeMap().get(typeReference.getName());
            } else {
                throw new UnresolvableTypeException("Invalid type when resolving union " + outputType.getClass().getSimpleName());
            }
        }
```
Hope it helps!

**Comment by kaqqao** [comment]
Are you on the latest version?

**Comment by tufgupmi** [comment]
No. I started to use this library from 0.9.9 and have to do some little modifications to adapt to my project. But i looked the current code and the bug is still there. 

The bug is that you cast to GraphQLObjectType when the class is GraphQLTypeReference, and it is that class because the type is used as return type in a GraphQLQuery.

**Comment by kaqqao** [comment]
The code you're referring to is correct. A type reference is never supposed to end up in the union (or anywhere else for that matter) once the schema is constructed. I believe the underlying issue has been fixed since 0.9.9.
What changes do you have that prevent you from updating? SPQR is designed for maximal customizability, so there's a good chance you do not need to fork it. Otherwise, if the use-case is legit for general usage, I can add more extension points.

**Comment by tufgupmi** [comment]
I will try to explain what changes i' have made:

1) I have classes with method that return Function<T,E> or Supplier<T>. This methods i annotated with @GraphQLIgnore. Despite this i got an Exception when the schema is build, So i have to introduce this change in PublicResolverBuilder (buildResolvers): I filter method annotated with @GraphQLIgnore. The exception was solved:

```
 return Arrays.stream(rawType.getMethods())
                **.filter(method -> !method.isAnnotationPresent(GraphQLIgnore.class))**
                .filter(method -> isPackageAcceptable(method, rawType, params.getBasePackages()))

```
2) I am also using graphql-braid (that too use graphql-spqr). graphql-braid is for a microservice that merges all the schemas that i have in diffetene microservices. In one of those microservices i have a LocalTime property. The graphql-braid sends that LocalTime property to the other microservice as a Map with ("hour","minute","second"...). For this i have to chnage the way temporalScalar are parsed in Scalars.class. I have to add this ti be able to parse this scalar. 

```

            @Override
            public Object parseValue(Object input) {
                try {
                    if (input instanceof String) {
                        return fromString.apply((String) input);
                    }
                    if (input instanceof Long) {
                        return fromDate.apply(Instant.ofEpochMilli((Long) input));
                    }
                    if (type.isInstance(input)) {
                        return input;
                    }else{
                        **if (input instanceof HashMap) {
                            HashMap map = (HashMap)input;
                            if (map.containsKey("hour")) {
                                return LocalTime.of((Integer)map.get("hour"), (Integer)map.get("minute"), (Integer)map.get("second"));
                            }
                        }**
                    }
                    throw valueParsingException(input, String.class, Long.class);
                } catch (Exception e) {
                    throw new CoercingParseValueException("Value " + input + " could not be parsed into a " + name);
                }
            }
```

I know is not an elegant solutions. But it's working!

**Comment by kaqqao** [comment]
1. That specific [issue](https://github.com/leangen/graphql-spqr/issues/298) was fixed in v0.10.1
2. You never ever need to touch the logic for handling any specific type. What you do instead is provide a custom `TypeMapper` that does what you want. You can fully customize temporal scalars or any other type this way.

I'll close this issue as it seems addressed, but feel free to reopen if it is still relevant.
